presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -15,7 +15,6 @@
 
 import com.facebook.presto.Session;
```

Comment:
Discussion:

> When doing stats arithmetics we expect that the symbol stats are normalized. I didn't catch an explicit failure there, but it looks like dealing on normalizing stats is safer.

Normalizing can also increase error and introduce noise. For instance you would expect join tree stats to be the same at root independently on join order. However, this is not the case due to normalization. Let have two joins, `J1, J2`.
Let's say filter factor for join `J1` is `1/10` and for `J2` `10` (multiplies rows).

If on input you have 10 rows with 10 distinct values (on join key `K`) then if you first join `J1` and then `J2` then for column `K` you will end up with just 1 distinct value (normalization after `J1`).
If you first use `J2` and then `J1` then `K` will still have 10 distinct values.

Therefore I would drop the change if it's not motivated by real use case.

Reply:
`JoinStatsRule` rule is a `SimpleStatsRule`, that does the normalization of the final stats anyway. Therefore when estimating multiple joins, the intermediate stats are normalized anyway.

This patch only adds normalization to the intermediate stats, and ensure that the stats are normalized before applying the `addJoinComplementStats` logic. This follows the logic we use in estimating filters. The stats are normalized at every step.